File: codes/csharp/array_every.cs
using System; // Console, Func<>, Action<>, Action
using System.Linq; // Enumerable
using System.Collections.Generic; // Dictionary<>, List<>, KeyValuePair<>, IEnumerable<>

class Program {
    static void Main(string[] Args) {
        dynamic IsNumeric = (Func<dynamic, bool>)((dynamic Anything) => (Anything is sbyte || Anything is byte || Anything is short || Anything is ushort || Anything is int || Anything is uint || Anything is long || Anything is ulong || Anything is float || Anything is double || Anything is decimal));

        string PrettyJsonStringify(dynamic Anything, string Indent = "    ") {
            int IndentLevel = 0;
            dynamic PrettyJsonStringifyInner = null;
            PrettyJsonStringifyInner = (Func<dynamic, string, string>)((dynamic AnythingInner, string IndentInner) => {
                if (AnythingInner == null) {
                    return "null";
                }
                if (AnythingInner is string) {
                    return "\"" + (string)AnythingInner + "\"";
                }
                if ((IsNumeric(AnythingInner) == true) || AnythingInner is bool) {
                    return AnythingInner.ToString().Replace(",", ".");
                }
                if (AnythingInner is List<dynamic>) {
                    IndentLevel += 1;
                    string Result = "[" + Environment.NewLine + string.Concat(Enumerable.Repeat(IndentInner, IndentLevel));
                    int ArrayItemIndex = 0;
                    foreach (dynamic ArrayItem in (List<dynamic>)AnythingInner) {
                        Result += PrettyJsonStringifyInner(ArrayItem, IndentInner);
                        if ((ArrayItemIndex + 1) != ((List<dynamic>)AnythingInner).Count) {
                            Result += "," + Environment.NewLine + string.Concat(Enumerable.Repeat(IndentInner, IndentLevel));
                        }
                        ArrayItemIndex += 1;
                    }
                    IndentLevel -= 1;
                    Result += Environment.NewLine + string.Concat(Enumerable.Repeat(IndentInner, IndentLevel)) + "]";
                    return Result;
                }
                if (AnythingInner is Dictionary<string, dynamic>) {
                    IndentLevel += 1;
                    string Result = "{" + Environment.NewLine + string.Concat(Enumerable.Repeat(IndentInner, IndentLevel));
                    int IterationIndex = 0;
                    foreach (KeyValuePair<string, dynamic> ObjectEntry in (Dictionary<string, dynamic>)AnythingInner) {
                        string ObjectKey = ObjectEntry.Key;
                        dynamic ObjectValue = ObjectEntry.Value;
                        Result += "\"" + ObjectKey + "\": " + PrettyJsonStringifyInner(ObjectValue, IndentInner);
                        if ((IterationIndex + 1) != ((Dictionary<string, dynamic>)AnythingInner).Count) {
                            Result += "," + Environment.NewLine + string.Concat(Enumerable.Repeat(IndentInner, IndentLevel));
                        }
                        IterationIndex += 1;
                    }
                    IndentLevel -= 1;
                    Result += Environment.NewLine + string.Concat(Enumerable.Repeat(IndentInner, IndentLevel)) + "}";
                    return Result;
                }
                return "null";
            });
            return PrettyJsonStringifyInner(Anything, Indent);
        }

        string PrettyArrayOfPrimitives(List<dynamic> AnArrayOfPrimitives) {
            string Result = "[";
            int ArrayItemIndex = 0;
            foreach (dynamic ArrayItem in AnArrayOfPrimitives) {
                if (((ArrayItem is string) == false) && (IsNumeric(ArrayItem) == false) && ((ArrayItem is bool) == false) && (ArrayItem != null)) continue;
                if (ArrayItem is string) Result += $"\"{ArrayItem}\"";
                if (IsNumeric(ArrayItem) == true) Result += ArrayItem.ToString().Replace(",", ".");
                if (ArrayItem is bool) Result += $"{ArrayItem.ToLower()}";
                if (ArrayItem == null) Result += "null";
                if ((ArrayItemIndex + 1) != AnArrayOfPrimitives.Count) Result += ", ";
                ArrayItemIndex += 1;
            }
            Result += "]";
            return Result;
        }

        bool ArrayEveryV1(Func<dynamic, dynamic, List<dynamic>, bool> CallbackFunction, List<dynamic> AnArray) {
            // JavaScript-like Array.every() function
            bool IsConditionMatch = false;
            int ArrayItemIndex = 0;
            foreach (dynamic ArrayItem in AnArray) {
                IsConditionMatch = (bool)CallbackFunction(ArrayItem, ArrayItemIndex, AnArray);
                if (IsConditionMatch == false) break;
                ArrayItemIndex += 1;
            }
            return IsConditionMatch;
        }

        bool ArrayEveryV2(Func<dynamic, dynamic, List<dynamic>, bool> CallbackFunction, List<dynamic> AnArray) {
            // JavaScript-like Array.every() function
            bool IsConditionMatch = false;
            int ArrayItemIndex = 0;
            foreach (dynamic ArrayItem in AnArray) {
                IsConditionMatch = (bool)CallbackFunction(ArrayItem, ArrayItemIndex, AnArray);
                if (IsConditionMatch == false) return IsConditionMatch;
                ArrayItemIndex += 1;
            }
            return IsConditionMatch;
        }

        bool ArrayEveryV3(Func<dynamic, dynamic, List<dynamic>, bool> CallbackFunction, List<dynamic> AnArray) {
            // JavaScript-like Array.every() function
            int ArrayItemIndex = 0;
            foreach (dynamic ArrayItem in AnArray) {
                bool IsConditionMatch = (bool)CallbackFunction(ArrayItem, ArrayItemIndex, AnArray);
                if (IsConditionMatch == false) return false;
                ArrayItemIndex += 1;
            }
            return true;
        }

        bool ArrayEveryV4(Func<dynamic, dynamic, List<dynamic>, bool> CallbackFunction, List<dynamic> AnArray) {
            // JavaScript-like Array.every() function
            int ArrayItemIndex = 0;
            foreach (dynamic ArrayItem in AnArray) {
                if ((bool)CallbackFunction(ArrayItem, ArrayItemIndex, AnArray) == false) return false;
                ArrayItemIndex += 1;
            }
            return true;
        }

        Console.WriteLine("\n// JavaScript-like Array.every() in C# List");

        dynamic Numbers = new List<dynamic>() {12, 34, 27, 23, 65, 93, 36, 87, 4, 254};
        Console.WriteLine($"Numbers: {PrettyArrayOfPrimitives(Numbers)}");

        dynamic IsAllNumberLessThan500;
        dynamic IsAllNumberMoreThan500;

        Console.WriteLine("// using JavaScript-like Array.every() function \"ArrayEveryV1\"");

        IsAllNumberLessThan500 = ArrayEveryV1((Func<dynamic, dynamic, List<dynamic>, bool>)((dynamic Number, dynamic ArrayItemIndex, List<dynamic> AnArray) => (Number < 500)), Numbers);
        Console.WriteLine($"is all number < 500: {IsAllNumberLessThan500}");
        // is all number < 500: True

        IsAllNumberMoreThan500 = ArrayEveryV1((Func<dynamic, dynamic, List<dynamic>, bool>)((dynamic Number, dynamic ArrayItemIndex, List<dynamic> AnArray) => (Number > 500)), Numbers);
        Console.WriteLine($"is all number > 500: {IsAllNumberMoreThan500}");
        // is all number > 500: False

        Console.WriteLine("// using JavaScript-like Array.every() function \"ArrayEveryV2\"");

        IsAllNumberLessThan500 = ArrayEveryV2((Func<dynamic, dynamic, List<dynamic>, bool>)((dynamic Number, dynamic ArrayItemIndex, List<dynamic> AnArray) => (Number < 500)), Numbers);
        Console.WriteLine($"is all number < 500: {IsAllNumberLessThan500}");
        // is all number < 500: True

        IsAllNumberMoreThan500 = ArrayEveryV2((Func<dynamic, dynamic, List<dynamic>, bool>)((dynamic Number, dynamic ArrayItemIndex, List<dynamic> AnArray) => (Number > 500)), Numbers);
        Console.WriteLine($"is all number > 500: {IsAllNumberMoreThan500}");
        // is all number > 500: False

        Console.WriteLine("// using JavaScript-like Array.every() function \"ArrayEveryV3\"");

        IsAllNumberLessThan500 = ArrayEveryV3((Func<dynamic, dynamic, List<dynamic>, bool>)((dynamic Number, dynamic ArrayItemIndex, List<dynamic> AnArray) => (Number < 500)), Numbers);
        Console.WriteLine($"is all number < 500: {IsAllNumberLessThan500}");
        // is all number < 500: True

        IsAllNumberMoreThan500 = ArrayEveryV3((Func<dynamic, dynamic, List<dynamic>, bool>)((dynamic Number, dynamic ArrayItemIndex, List<dynamic> AnArray) => (Number > 500)), Numbers);
        Console.WriteLine($"is all number > 500: {IsAllNumberMoreThan500}");
        // is all number > 500: False

        Console.WriteLine("// using JavaScript-like Array.every() function \"ArrayEveryV4\"");

        IsAllNumberLessThan500 = ArrayEveryV4((Func<dynamic, dynamic, List<dynamic>, bool>)((dynamic Number, dynamic ArrayItemIndex, List<dynamic> AnArray) => (Number < 500)), Numbers);
        Console.WriteLine($"is all number < 500: {IsAllNumberLessThan500}");
        // is all number < 500: True

        IsAllNumberMoreThan500 = ArrayEveryV4((Func<dynamic, dynamic, List<dynamic>, bool>)((dynamic Number, dynamic ArrayItemIndex, List<dynamic> AnArray) => (Number > 500)), Numbers);
        Console.WriteLine($"is all number > 500: {IsAllNumberMoreThan500}");
        // is all number > 500: False

        Console.WriteLine("// using C# Array.every() built-in method \"IEnumerable.All()\"");

        IsAllNumberLessThan500 = ((IEnumerable<dynamic>)Numbers).All((Func<dynamic, bool>)((dynamic Number) => (Number < 500)));
        Console.WriteLine($"is all number < 500: {IsAllNumberLessThan500}");
        // is all number < 500: True

        IsAllNumberMoreThan500 = ((IEnumerable<dynamic>)Numbers).All((Func<dynamic, bool>)((dynamic Number) => (Number > 500)));
        Console.WriteLine($"is all number > 500: {IsAllNumberMoreThan500}");
        // is all number > 500: False

        Console.WriteLine("\n// JavaScript-like Array.every() in C# List of Dictionaries");

        dynamic Products = new List<dynamic>() {
            new Dictionary<string, dynamic>() {
                {"code", "pasta"},
                {"price", 321}
            },
            new Dictionary<string, dynamic>() {
                {"code", "bubble_gum"},
                {"price", 233}
            },
            new Dictionary<string, dynamic>() {
                {"code", "potato_chips"},
                {"price", 5}
            },
            new Dictionary<string, dynamic>() {
                {"code", "towel"},
                {"price", 499}
            },
        };
        Console.WriteLine($"Products: {PrettyJsonStringify(Products)}");

        dynamic IsAllProductPriceLessThan500;
        dynamic IsAllProductPriceMoreThan500;

        Console.WriteLine("// using JavaScript-like Array.every() function \"ArrayEveryV1\"");

        IsAllProductPriceLessThan500 = ArrayEveryV1((Func<dynamic, dynamic, List<dynamic>, bool>)((dynamic Product, dynamic ArrayItemIndex, List<dynamic> AnArray) => (Product["price"] < 500)), Products);
        Console.WriteLine($"is all product price < 500: {IsAllProductPriceLessThan500}");
        // is all product price < 500: True

        IsAllProductPriceMoreThan500 = ArrayEveryV1((Func<dynamic, dynamic, List<dynamic>, bool>)((dynamic Product, dynamic ArrayItemIndex, List<dynamic> AnArray) => (Product["price"] > 500)), Products);
        Console.WriteLine($"is all product price > 500: {IsAllProductPriceMoreThan500}");
        // is all product price > 500: False

        Console.WriteLine("// using JavaScript-like Array.every() function \"ArrayEveryV2\"");

        IsAllProductPriceLessThan500 = ArrayEveryV2((Func<dynamic, dynamic, List<dynamic>, bool>)((dynamic Product, dynamic ArrayItemIndex, List<dynamic> AnArray) => (Product["price"] < 500)), Products);
        Console.WriteLine($"is all product price < 500: {IsAllProductPriceLessThan500}");
        // is all product price < 500: True

        IsAllProductPriceMoreThan500 = ArrayEveryV2((Func<dynamic, dynamic, List<dynamic>, bool>)((dynamic Product, dynamic ArrayItemIndex, List<dynamic> AnArray) => (Product["price"] > 500)), Products);
        Console.WriteLine($"is all product price > 500: {IsAllProductPriceMoreThan500}");
        // is all product price > 500: False

        Console.WriteLine("// using JavaScript-like Array.every() function \"ArrayEveryV3\"");

        IsAllProductPriceLessThan500 = ArrayEveryV3((Func<dynamic, dynamic, List<dynamic>, bool>)((dynamic Product, dynamic ArrayItemIndex, List<dynamic> AnArray) => (Product["price"] < 500)), Products);
        Console.WriteLine($"is all product price < 500: {IsAllProductPriceLessThan500}");
        // is all product price < 500: True

        IsAllProductPriceMoreThan500 = ArrayEveryV3((Func<dynamic, dynamic, List<dynamic>, bool>)((dynamic Product, dynamic ArrayItemIndex, List<dynamic> AnArray) => (Product["price"] > 500)), Products);
        Console.WriteLine($"is all product price > 500: {IsAllProductPriceMoreThan500}");
        // is all product price > 500: False

        Console.WriteLine("// using JavaScript-like Array.every() function \"ArrayEveryV4\"");

        IsAllProductPriceLessThan500 = ArrayEveryV4((Func<dynamic, dynamic, List<dynamic>, bool>)((dynamic Product, dynamic ArrayItemIndex, List<dynamic> AnArray) => (Product["price"] < 500)), Products);
        Console.WriteLine($"is all product price < 500: {IsAllProductPriceLessThan500}");
        // is all product price < 500: True

        IsAllProductPriceMoreThan500 = ArrayEveryV4((Func<dynamic, dynamic, List<dynamic>, bool>)((dynamic Product, dynamic ArrayItemIndex, List<dynamic> AnArray) => (Product["price"] > 500)), Products);
        Console.WriteLine($"is all product price > 500: {IsAllProductPriceMoreThan500}");
        // is all product price > 500: False

        Console.WriteLine("// using C# Array.every() built-in method \"IEnumerable.All()\"");

        IsAllProductPriceLessThan500 = ((IEnumerable<dynamic>)Products).All((Func<dynamic, bool>)((dynamic Product) => (Product["price"] < 500)));
        Console.WriteLine($"is all product price < 500: {IsAllProductPriceLessThan500}");
        // is all product price < 500: True

        IsAllProductPriceMoreThan500 = ((IEnumerable<dynamic>)Products).All((Func<dynamic, bool>)((dynamic Product) => (Product["price"] > 500)));
        Console.WriteLine($"is all product price > 500: {IsAllProductPriceMoreThan500}");
        // is all product price > 500: False
    }
}

File: codes/dart/array_every.dart
import "dart:convert";

void main() {
    dynamic prettyArrayOfPrimitives(dynamic anything) {
        final jsonEncoder = new JsonEncoder();
        return jsonEncoder.convert(anything).split(',').join(', ');
    }

    dynamic prettyJsonStringify(dynamic anything) {
        final jsonEncoder = new JsonEncoder.withIndent("    ");
        return jsonEncoder.convert(anything);
    }

    dynamic arrayEveryV1(dynamic callbackFunction, dynamic anArray) {
        // JavaScript-like Array.every() function
        dynamic isConditionMatch = false;
        for (final entryItem in anArray.asMap().entries) {
            final arrayItemIndex = entryItem.key;
            final arrayItem = entryItem.value;
            isConditionMatch = callbackFunction(arrayItem, arrayItemIndex, anArray);
            if (isConditionMatch == false) break;
        }
        return isConditionMatch;
    }

    dynamic arrayEveryV2(dynamic callbackFunction, dynamic anArray) {
        // JavaScript-like Array.every() function
        dynamic isConditionMatch = false;
        for (final entryItem in anArray.asMap().entries) {
            final arrayItemIndex = entryItem.key;
            final arrayItem = entryItem.value;
            isConditionMatch = callbackFunction(arrayItem, arrayItemIndex, anArray);
            if (isConditionMatch == false) return isConditionMatch;
        }
        return isConditionMatch;
    }

    dynamic arrayEveryV3(dynamic callbackFunction, dynamic anArray) {
        // JavaScript-like Array.every() function
        for (final entryItem in anArray.asMap().entries) {
            final arrayItemIndex = entryItem.key;
            final arrayItem = entryItem.value;
            dynamic isConditionMatch = callbackFunction(arrayItem, arrayItemIndex, anArray);
            if (isConditionMatch == false) return false;
        }
        return true;
    }

    dynamic arrayEveryV4(dynamic callbackFunction, dynamic anArray) {
        // JavaScript-like Array.every() function
        for (final entryItem in anArray.asMap().entries) {
            final arrayItemIndex = entryItem.key;
            final arrayItem = entryItem.value;
            if (callbackFunction(arrayItem, arrayItemIndex, anArray) == false) return false;
        }
        return true;
    }

    print("\n// JavaScript-like Array.every() in Dart List");

    final numbers = <dynamic>[12, 34, 27, 23, 65, 93, 36, 87, 4, 254];
    print("numbers: ${prettyArrayOfPrimitives(numbers)}");

    dynamic isAllNumberLessThan500;
    dynamic isAllNumberMoreThan500;

    print("// using JavaScript-like Array.every() function \"arrayEveryV1\"");

    isAllNumberLessThan500 = arrayEveryV1((dynamic number, dynamic _, dynamic __) => (number < 500), numbers);
    print("is all number < 500: ${isAllNumberLessThan500}");
    // is all number < 500: true

    isAllNumberMoreThan500 = arrayEveryV1((dynamic number, dynamic _, dynamic __) => (number > 500), numbers);
    print("is all number > 500: ${isAllNumberMoreThan500}");
    // is all number > 500: false

    print("// using JavaScript-like Array.every() function \"arrayEveryV2\"");

    isAllNumberLessThan500 = arrayEveryV2((dynamic number, dynamic _, dynamic __) => (number < 500), numbers);
    print("is all number < 500: ${isAllNumberLessThan500}");
    // is all number < 500: true

    isAllNumberMoreThan500 = arrayEveryV2((dynamic number, dynamic _, dynamic __) => (number > 500), numbers);
    print("is all number > 500: ${isAllNumberMoreThan500}");
    // is all number > 500: false

    print("// using JavaScript-like Array.every() function \"arrayEveryV3\"");

    isAllNumberLessThan500 = arrayEveryV3((dynamic number, dynamic _, dynamic __) => (number < 500), numbers);
    print("is all number < 500: ${isAllNumberLessThan500}");
    // is all number < 500: true

    isAllNumberMoreThan500 = arrayEveryV3((dynamic number, dynamic _, dynamic __) => (number > 500), numbers);
    print("is all number > 500: ${isAllNumberMoreThan500}");
    // is all number > 500: false

    print("// using JavaScript-like Array.every() function \"arrayEveryV4\"");

    isAllNumberLessThan500 = arrayEveryV4((dynamic number, dynamic _, dynamic __) => (number < 500), numbers);
    print("is all number < 500: ${isAllNumberLessThan500}");
    // is all number < 500: true

    isAllNumberMoreThan500 = arrayEveryV4((dynamic number, dynamic _, dynamic __) => (number > 500), numbers);
    print("is all number > 500: ${isAllNumberMoreThan500}");
    // is all number > 500: false

    print("// using Dart Array.every() built-in method \"Array.every()\"");

    isAllNumberLessThan500 = numbers.every((dynamic number) => (number < 500));
    print("is all number < 500: ${isAllNumberLessThan500}");
    // is all number < 500: true

    isAllNumberMoreThan500 = numbers.every((dynamic number) => (number > 500));
    print("is all number > 500: ${isAllNumberMoreThan500}");
    // is all number > 500: false

    print("\n// JavaScript-like Array.every() in Dart List of Maps");

    final products = <dynamic>[
        <String, dynamic>{
            "code": "pasta",
            "price": 321
        },
        <String, dynamic>{
            "code": "bubble_gum",
            "price": 233
        },
        <String, dynamic>{
            "code": "potato_chips",
            "price": 5
        },
        <String, dynamic>{
            "code": "towel",
            "price": 499
        }
    ];
    print("products: ${prettyJsonStringify(products)}");

    dynamic isAllProductPriceLessThan500;
    dynamic isAllProductPriceMoreThan500;

    print("// using JavaScript-like Array.every() function \"arrayEveryV1\"");

    isAllProductPriceLessThan500 = arrayEveryV1((dynamic product, dynamic _, dynamic __) => (product["price"] < 500), products);
    print("is all product price < 500: ${isAllProductPriceLessThan500}");
    // is all product price < 500: true

    isAllProductPriceMoreThan500 = arrayEveryV1((dynamic product, dynamic _, dynamic __) => (product["price"] > 500), products);
    print("is all product price > 500: ${isAllProductPriceMoreThan500}");
    // is all product price > 500: false

    print("// using JavaScript-like Array.every() function \"arrayEveryV2\"");

    isAllProductPriceLessThan500 = arrayEveryV2((dynamic product, dynamic _, dynamic __) => (product["price"] < 500), products);
    print("is all product price < 500: ${isAllProductPriceLessThan500}");
    // is all product price < 500: true

    isAllProductPriceMoreThan500 = arrayEveryV2((dynamic product, dynamic _, dynamic __) => (product["price"] > 500), products);
    print("is all product price > 500: ${isAllProductPriceMoreThan500}");
    // is all product price > 500: false

    print("// using JavaScript-like Array.every() function \"arrayEveryV3\"");

    isAllProductPriceLessThan500 = arrayEveryV3((dynamic product, dynamic _, dynamic __) => (product["price"] < 500), products);
    print("is all product price < 500: ${isAllProductPriceLessThan500}");
    // is all product price < 500: true

    isAllProductPriceMoreThan500 = arrayEveryV3((dynamic product, dynamic _, dynamic __) => (product["price"] > 500), products);
    print("is all product price > 500: ${isAllProductPriceMoreThan500}");
    // is all product price > 500: false

    print("// using JavaScript-like Array.every() function \"arrayEveryV4\"");

    isAllProductPriceLessThan500 = arrayEveryV4((dynamic product, dynamic _, dynamic __) => (product["price"] < 500), products);
    print("is all product price < 500: ${isAllProductPriceLessThan500}");
    // is all product price < 500: true

    isAllProductPriceMoreThan500 = arrayEveryV4((dynamic product, dynamic _, dynamic __) => (product["price"] > 500), products);
    print("is all product price > 500: ${isAllProductPriceMoreThan500}");
    // is all product price > 500: false

    print("// using Dart Array.every() built-in method \"Array.every()\"");

    isAllProductPriceLessThan500 = products.every((dynamic product) => (product["price"] < 500));
    print("is all product price < 500: ${isAllProductPriceLessThan500}");
    // is all product price < 500: true

    isAllProductPriceMoreThan500 = products.every((dynamic product) => (product["price"] > 500));
    print("is all product price > 500: ${isAllProductPriceMoreThan500}");
    // is all product price > 500: false
}
